packages/runtime/test-runtime-utils/src/mocksForRebasing.ts

```diff
@@ -0,0 +1,155 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { v4 as uuid } from "uuid";
+import { ISequencedDocumentMessage, MessageType } from "@fluidframework/protocol-definitions";
+import {
+	MockContainerRuntime,
+	MockContainerRuntimeFactory,
+	MockFluidDataStoreRuntime,
+} from "./mocks";
+
+/**
+ * Specialized implementation of MockContainerRuntime for testing op rebasing, when the runtime will resend
+ * ops to the datastores and all ops within the same batch will have the same sequence number.
+ */
+export class MockContainerRuntimeForRebasing extends MockContainerRuntime {
+	private readonly currentBatch: ITrackableMessage[] = [];
+
+	constructor(
+		dataStoreRuntime: MockFluidDataStoreRuntime,
+		factory: MockContainerRuntimeFactoryForRebasing,
+		overrides?: { minimumSequenceNumber?: number },
+	) {
+		super(dataStoreRuntime, factory, overrides);
+	}
+
+	public process(message: ISequencedDocumentMessage) {
+		// Processing ops will happen in a separate JS turn, so by then, we'd increase
+		// the sequence number and flush the current batch.
+		this.clientSequenceNumber++;
+		this.currentBatch.splice(0);
+
+		super.process(message);
+	}
+
+	public submit(messageContent: any, localOpMetadata: unknown) {
+		const message = {
+			content: messageContent,
+			localOpMetadata,
+			opId: uuid(),
+			timesSubmitted: 0,
+		};
+		this.submitInternal(message);
+		this.currentBatch.push(message);
+
+		// Messages in the same batch will have the same clientSequenceNumber
+		return this.clientSequenceNumber;
+	}
+
+	private submitInternal(message: ITrackableMessage) {
+		message.timesSubmitted++;
+
+		const metadata = { opId: message.opId, timesSubmitted: message.timesSubmitted };
+		this.factory.pushMessage({
+			clientId: this.clientId,
+			clientSequenceNumber: this.clientSequenceNumber,
+			contents: message.content,
+			referenceSequenceNumber: this.referenceSequenceNumber,
+			type: MessageType.Operation,
+			metadata,
+		});
+		this.addPendingMessage(
+			message.content,
+			{ ...(message.localOpMetadata as object), ...metadata },
+			this.clientSequenceNumber,
+		);
+	}
+
+	public rebase() {
+		this.currentBatch.forEach((message) => this.submitInternal(message));
```

Comment:
I'd think that the purpose of this mock is to roundtrip messages through DDSs on resubmit. But here we do simple resubmit....

Reply:
Did some more digging (and a repro with the Sequence DDS which proved we're sending the messages twice) and you are correct. Apparently, the underlying mocks don't support this yet 😞. Check out `MockFluidDataStoreRuntime`.